device/core/src/module_client.ts

```diff
@@ -361,13 +364,25 @@ export class ModuleClient extends InternalClient {
    *     - IOTEDGE_AUTHSCHEME           Authentication scheme to use; must be "sasToken"
    *
    * @param transportCtor Transport protocol used to connect to IoT hub.
-   * @param [callback]    Optional callback to invoke when the ModuleClient has been constructured or if an
+   * @param modelId       The modelId to include in the username on connect as per the Azure IoT Plug and Play requirements.
+   * @param [callback]    Optional callback to invoke when the ModuleClient has been constructed or if an
    *                      error occurs while creating the client.
    * @returns {Promise<ModuleClient> | void} Promise if no callback function was passed, void otherwise.
    */
-  static fromEnvironment(transportCtor: any, callback: Callback<ModuleClient>): void;
   static fromEnvironment(transportCtor: any): Promise<ModuleClient>;
-  static fromEnvironment(transportCtor: any, callback?: Callback<ModuleClient>): Promise<ModuleClient> | void {
+  static fromEnvironment(transportCtor: any, modelId: string): Promise<ModuleClient>;
+  static fromEnvironment(transportCtor: any, callback: Callback<ModuleClient>): void;
+  static fromEnvironment(transportCtor: any, modelId: string, callback: Callback<ModuleClient>): void;
+  static fromEnvironment(transportCtor: any, callbackOrModelId?: Callback<ModuleClient> | string, callback?: Callback<ModuleClient>): Promise<ModuleClient> | void {
+    if (callbackOrModelId && typeof(callbackOrModelId) !== 'function' && typeof(callbackOrModelId) !== 'string') {
+      throw new TypeError('The second argument to fromEnvironment must be a callback or a modelId string.');
+    }
```

Comment:
I forgot how much I HATE HATE HATE function overloading in node.  Out of all the code in this review, this is the most fragile looking.  (No insult intended, this is just the way we do it, and it sucks.)  I recommend adding unit tests for this to test all of the various new valid and invalid cases this code adds.

Reply:
Function overloading is made much worse in our situation because we have to retain callback functionality. IMO, it wouldn't be that bad if we got rid of the callbacks. Hopefully we don't have to deal with this in Track 2.

Yeah, there needs to be tests for `fromEnvironment`. Since testing `fromEnviroment` seems much more involved than the other factory methods, I was planning on doing it in a later PR (but before we merge the branch to master). I have it on a list of things I need to address along with writing E2E tests.

If you're not comfortable approving this PR without those tests, I can go ahead and work on them now.

Reply:
Yeah, I'm OK without the tests.  And I agree with what you say.